pkg/controller/mysqlcluster/internal/cleaner/pvc.go

```diff
@@ -0,0 +1,189 @@
+	messagePvcNotDeleted = "delete Claim %s in StatefulSet %s failed"
+)
+
+var log = logf.Log.WithName(controllerName)
+
+// PvcCleaner represents an object to clean Pvcs of a MysqlCluster
+type PvcCleaner struct {
+	cluster *wrapcluster.MysqlCluster
+	opt     *options.Options
+	sts     *apps.StatefulSet
+}
+
+// NewPvcCleaner returns a new PVC cleaner object
+func NewPvcCleaner(cluster *api.MysqlCluster, opt *options.Options) *PvcCleaner {
+	obj := &apps.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.GetNameForResource(api.StatefulSet),
+			Namespace: cluster.Namespace,
+		},
+	}
+	return &PvcCleaner{
+		cluster: wrapcluster.NewMysqlClusterWrapper(cluster),
+		opt:     opt,
+		sts:     obj,
+	}
+}
+
+// Run performs cleanup of orphaned pvcs in a Mysql cluster
+func (p *PvcCleaner) Run(ctx context.Context, c client.Client, scheme *runtime.Scheme, recorder record.EventRecorder) error {
+	sts := p.sts
+	stsMeta := sts.ObjectMeta
+
+	key := types.NamespacedName{Name: stsMeta.GetName(), Namespace: stsMeta.GetNamespace()}
+
+	err := c.Get(ctx, key, sts)
+	if err != nil {
+		return err
+	}
+
+	if sts.DeletionTimestamp != nil {
+		log.V(4).Info("being deleted, no action", "Statefulset", key)
+		return nil
+	}
+
+	if len(sts.Spec.VolumeClaimTemplates) == 0 {
+		log.V(4).Info("ignored, no volume claims", "Statefulset", key)
+		return nil
+	}
+
+	replicas := sts.Spec.Replicas
+
+	// Find any pvcs with higher ordinal than replicas and delete them
+	claims, err := p.getClaims(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	keys := getKeys(claims)
+	sort.Sort(sort.Reverse(sort.IntSlice(keys)))
+
+	for _, k := range keys {
+		if int32(k) >= *replicas {
+			log.Info("cleaning up", "pvc", claims[k])
+			if err := deleteClaim(ctx, c, recorder, sts, claims[k]); err != nil {
+				log.Error(err, "deleting claim")
+			}
+		}
+	}
+	return nil
+}
+
+func deleteClaim(ctx context.Context, c client.Client, recorder record.EventRecorder, sts *apps.StatefulSet, pvc *core.PersistentVolumeClaim) error {
+	err := c.Delete(ctx, pvc)
+	if err != nil {
+		recorder.Event(sts, core.EventTypeWarning, deleteFail, fmt.Sprintf(messagePvcNotDeleted, pvc.Name, sts.Name))
+		return err
+	}
+
+	recorder.Event(sts, core.EventTypeNormal, deleteSuccess, fmt.Sprintf(messagePvcDeleted, pvc.Name, sts.Name))
+	return nil
+}
+func (p *PvcCleaner) getClaims(ctx context.Context, c client.Client) (map[int]*core.PersistentVolumeClaim, error) {
+	sts := p.sts
+	stsMeta := sts.ObjectMeta
+	pvcs := &core.PersistentVolumeClaimList{}
+	lo := &client.ListOptions{
+		Namespace: stsMeta.GetNamespace(),
+		Raw: &metav1.ListOptions{
+			LabelSelector: getLabelString(p.cluster.GetLabels()),
+		}}
+	err := c.List(ctx, lo, pvcs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return byOrdinal(pvcs), nil
+}
+
+func getLabelString(labels map[string]string) string {
+	return fmt.Sprintf("app=%s,mysql_cluster=%s", labels["app"], labels["mysql_cluster"])
```

Comment:
`k8s.io/apimachinery/pkg/labels` provides functions to convert labels to selectors. https://godoc.org/k8s.io/apimachinery/pkg/labels. We should use those instead.